File: Stage16/Python/Problem28279.py
from collections import deque
import sys
input = sys.stdin.readline

def main():
    N = int(input())
    dq = deque()
    
    for _ in range(N):
        cmd = (input().split())
        
        if cmd[0] == '1':
            dq.appendleft(int(cmd[1]))
        elif cmd[0] == '2':
            dq.append(int(cmd[1]))
        elif cmd[0] == '3':
            print(dq.popleft() if dq else -1)
        elif cmd[0] == '4':
            print(dq.pop() if dq else -1)
        elif cmd[0] == '5':
            print(len(dq))
        elif cmd[0] == '6':
            print(0 if dq else 1)
        elif cmd[0] == '7':
            print(dq[0] if dq else -1)
        elif cmd[0] == '8':
            print(dq[-1] if dq else -1)
    
if __name__ == "__main__":
    main()

File: Stage16/Java/Problem28279.java
package Stage16.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Problem28279 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine());
        Deque<Integer> dq = new ArrayDeque<>();

        while (N-- > 0) {
            String[] input = br.readLine().split(" ");
            int command = Integer.parseInt(input[0]);

            switch (command) {
                case 1:
                    int X1 = Integer.parseInt(input[1]);
                    dq.addFirst(X1);
                    break;
                case 2:
                    int X2 = Integer.parseInt(input[1]);
                    dq.addLast(X2);
                    break;
                case 3:
                    sb.append(dq.isEmpty() ? -1 : dq.pollFirst()).append("\n");
                    break;
                case 4:
                    sb.append(dq.isEmpty() ? -1 : dq.pollLast()).append("\n");
                    break;
                case 5:
                    sb.append(dq.size()).append("\n");
                    break;
                case 6:
                    sb.append(dq.isEmpty() ? 1 : 0).append("\n");
                    break;
                case 7:
                    sb.append(dq.isEmpty() ? -1 : dq.peekFirst()).append("\n");
                    break;
                case 8:
                    sb.append(dq.isEmpty() ? -1 : dq.peekLast()).append("\n");
                    break;
            }
        }

        System.out.println(sb);
        br.close();
    }
}
